GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs

```diff
@@ -1094,14 +1099,25 @@ private void UpdatePlaceholders()
 
                 this.blobSizes.Flush();
 
-                // Waiting for the UpdateOrDeleteFilePlaceholder to fill the folderPlaceholdersToKeep before trying to remove folder placeholders
-                // so that we don't try to delete a folder placeholder that has file placeholders and just fails
-                foreach (PlaceholderListDatabase.PlaceholderData folderPlaceholder in placeholderListCopy.Where(x => x.IsFolder).OrderByDescending(x => x.Path))
-                {
-                    this.TryRemoveFolderPlaceholder(folderPlaceholder, updatedPlaceholderList, folderPlaceholdersToKeep);
+                using (BlobSizes.BlobSizesConnection blobSizesConnection = this.blobSizes.CreateConnection())
+                {
+                    // Waiting for the UpdateOrDeleteFilePlaceholder to fill the folderPlaceholdersToKeep before trying to remove folder placeholders
+                    // so that we don't try to delete a folder placeholder that has file placeholders and just fails
+                    IEnumerable<PlaceholderListDatabase.PlaceholderData> folderPlaceholderData = placeholderListCopy.Where(x => x.IsFolder);
+
+                    HashSet<string> folderPlaceholders = new HashSet<string>(folderPlaceholderData.Select(x => x.Path), StringComparer.OrdinalIgnoreCase);
+                    foreach (PlaceholderListDatabase.PlaceholderData folderPlaceholder in folderPlaceholderData.OrderByDescending(x => x.Path))
+                    {
+                        if (GVFSPlatform.Instance.FileSystem.EnumerationExpandsDirectories && folderPlaceholder.IsExpandedFolder)
+                        {
+                            this.ReExpandFolder(blobSizesConnection, folderPlaceholder.Path, updatedPlaceholderList, folderPlaceholders, folderPlaceholdersToKeep);
+                        }
+
```

Comment:
Are we missing an else here?

Reply:
There's no else because we still want to attempt to remove empty placeholders on Mac.

The call to `ReExpandFolder` happens first to (potentially) add more entries to `folderPlaceholdersToKeep`.

Let me know if you think a comment would be useful here.

Reply:
Yea a comment would be helpful, I think. And a functional test that fails if we add that else, because it's so very tempting to do so.